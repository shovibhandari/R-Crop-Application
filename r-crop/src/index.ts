export {RCrop as Cropper} from './r-crop';
export {RCrop as default} from './r-crop';
export type {RCropProps, RCropElement} from './r-crop';
