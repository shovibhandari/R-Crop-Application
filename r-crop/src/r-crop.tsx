import React, {useCallback, useEffect, useRef} from 'react';
import Cropper from 'cropperjs';
import {cleanImageProps} from './utils';

const REQUIRED_IMAGE_STYLES = {opacity: 0, maxWidth: '100%'};

interface RCropElement extends HTMLImageElement {
    cropper: Cropper;
}

type RCropRef =
    | ((instance: HTMLImageElement | RCropElement | null) => void)
    | React.MutableRefObject<HTMLImageElement | RCropElement | null>
    | null;

interface RCropDefaultOptions {
    scaleX?: number;
    scaleY?: number;
    enable?: boolean;
    zoomTo?: number;
    rotateTo?: number;
}

interface RCropProps
    extends RCropDefaultOptions,
        Cropper.Options<HTMLImageElement>,
        Omit<
            React.HTMLProps<HTMLImageElement>,
            'data' | 'ref' | 'crossOrigin'
        > {
    crossOrigin?: '' | 'anonymous' | 'use-credentials' | undefined;
    on?: (
        eventName: string,
        callback: () => void | Promise<void>,
    ) => void | Promise<void>;
    onInitialized?: (instance: Cropper) => void | Promise<void>;
}

const applyDefaultOptions = (
    cropper: Cropper,
    options: RCropDefaultOptions = {},
): void => {
    const {
        enable = true,
        scaleX = 1,
        scaleY = 1,
        zoomTo = 0,
        rotateTo,
    } = options;
    enable ? cropper.enable() : cropper.disable();
    cropper.scaleX(scaleX);
    cropper.scaleY(scaleY);
    rotateTo !== undefined && cropper.rotateTo(rotateTo);
    zoomTo > 0 && cropper.zoomTo(zoomTo);
};

const useCombinedRefs = (
    ...refs: RCropRef[]
): React.RefObject<RCropElement> => {
    const targetRef = useRef<RCropElement>(null);

    React.useEffect(() => {
        refs.forEach((ref) => {
            if (!ref) return;

            if (typeof ref === 'function') {
                ref(targetRef.current);
            } else {
                ref.current = targetRef.current;
            }
        });
    }, [refs]);

    return targetRef;
};

const RCrop = React.forwardRef<RCropElement | HTMLImageElement, RCropProps>(
    (props, ref) => {
        const {
            dragMode = 'crop',
            src,
            style,
            className,
            crossOrigin,
            scaleX,
            scaleY,
            enable,
            zoomTo,
            rotateTo,
            alt = 'picture',
            ready,
            onInitialized,
            ...rest
        } = props;

        const defaultOptions: RCropDefaultOptions = {
            scaleY,
            scaleX,
            enable,
            zoomTo,
            rotateTo,
        };

        const innerRef = useRef<HTMLImageElement>(null);
        const combinedRef = useCombinedRefs(ref, innerRef);

        const handleZoomTo = useCallback(() => {
            if (combinedRef.current?.cropper && typeof zoomTo === 'number') {
                combinedRef.current.cropper.zoomTo(zoomTo);
            }
        }, [zoomTo]);

        const handleSrcChange = useCallback(() => {
            if (combinedRef.current?.cropper && typeof src !== 'undefined') {
                combinedRef.current.cropper.reset().clear().replace(src);
            }
        }, [src]);

        const initializeCropper = useCallback(() => {
            if (combinedRef.current !== null) {
                const cropper = new Cropper(combinedRef.current, {
                    dragMode,
                    ...rest,
                    ready: (e) => {
                        if (e.currentTarget !== null) {
                            applyDefaultOptions(
                                e.currentTarget.cropper,
                                defaultOptions,
                            );
                        }
                        ready && ready(e);
                    },
                });
                onInitialized && onInitialized(cropper);
            }
        }, [combinedRef, dragMode, rest, ready, onInitialized, defaultOptions]);

        useEffect(handleZoomTo, [handleZoomTo]);
        useEffect(handleSrcChange, [handleSrcChange]);
        useEffect(initializeCropper, [initializeCropper]);

        const imageProps = cleanImageProps({...rest, crossOrigin, src, alt});

        return (
            <div style={style} className={className}>
                <img
                    {...imageProps}
                    style={REQUIRED_IMAGE_STYLES}
                    ref={combinedRef}
                />
            </div>
        );
    },
);

export {RCrop, RCropProps, RCropElement, applyDefaultOptions};
