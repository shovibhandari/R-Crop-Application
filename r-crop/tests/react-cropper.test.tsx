import React from 'react';
import {create} from 'react-test-renderer';
import {render, waitFor} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {RCrop as RCropper, applyDefaultOptions} from '../src/r-crop';

const image =
    'https://cdn.pixabay.com/photo/2023/07/10/06/52/trail-8117575_1280.jpg';
const newImage =
    'https://cdn.pixabay.com/photo/2016/11/14/04/45/elephant-1822636_1280.jpg';

describe('R crop render tests', () => {
    test('R crop snapshot', () => {
        const ref = React.createRef<HTMLImageElement>();
        const component = create(<RCropper src={image} ref={ref} />).toJSON();
        expect(component).toMatchSnapshot();
    });

    test('renders cropper and calls onInitialized', async () => {
        const onInitialized = jest.fn();
        const ref = React.createRef<HTMLImageElement>();
        render(
            <RCropper src={image} onInitialized={onInitialized} ref={ref} />,
        );
        await waitFor(() => expect(onInitialized).toHaveBeenCalledTimes(1));
    });

    test('cropper re-renders on src change', async () => {
        const ref = React.createRef<HTMLImageElement>();
        const {rerender, findByAltText} = render(
            <RCropper src={image} ref={ref} />,
        );
        const imageTag = await findByAltText('picture');
        expect(imageTag).toHaveAttribute('src', image);
        rerender(<RCropper src={newImage} ref={ref} />);
        expect(imageTag).toHaveAttribute('src', newImage);
    });

    test('renders cropper without ref and calls onInitialized', async () => {
        const onInitialized = jest.fn();
        render(<RCropper src={image} onInitialized={onInitialized} />);
        await waitFor(() => expect(onInitialized).toHaveBeenCalledTimes(1));
    });

    test('renders cropper with functional ref - method invoked', async () => {
        const onInitialized = jest.fn();
        const ref = jest.fn();
        render(
            <RCropper src={image} onInitialized={onInitialized} ref={ref} />,
        );
        await waitFor(() => expect(onInitialized).toHaveBeenCalledTimes(1));
        await waitFor(() => expect(ref).toHaveBeenCalled());
    });

    test('renders cropper with functional ref and check ref exists', async () => {
        const onInitialized = jest.fn();
        let ref: HTMLImageElement;
        const {rerender} = render(
            <RCropper
                src={image}
                onInitialized={onInitialized}
                ref={(cropperRef) => (ref = cropperRef)}
            />,
        );
        await waitFor(() => expect(onInitialized).toHaveBeenCalledTimes(1));
        expect(ref.src).toEqual(image);
        rerender(<RCropper src={newImage} />);
        expect(ref.src).toEqual(newImage);
    });

    test('renders cropper with zoomTo prop', async () => {
        const onInitialized = jest.fn();
        render(
            <RCropper
                src={newImage}
                onInitialized={onInitialized}
                zoomTo={1}
            />,
        );
        await waitFor(() => expect(onInitialized).toHaveBeenCalledTimes(1));
    });
});

describe('Test RCropper Methods', () => {
    const enable = jest.fn();
    const scaleX = jest.fn();
    const scaleY = jest.fn();
    const rotateTo = jest.fn();
    const zoomTo = jest.fn();
    const disable = jest.fn();

    let cropper;
    beforeAll(() => {
        cropper = {enable, disable, scaleX, scaleY, rotateTo, zoomTo};
    });

    test('Test applyDefaultOptions with default params', () => {
        applyDefaultOptions(cropper);
        expect(enable).toHaveBeenCalledTimes(1);
        expect(disable).toHaveBeenCalledTimes(0);

        expect(scaleX).toHaveBeenCalledWith(1);
        expect(scaleY).toHaveBeenCalledWith(1);
        expect(scaleX).toHaveBeenCalledTimes(1);
        expect(scaleY).toHaveBeenCalledTimes(1);

        expect(zoomTo).toHaveBeenCalledTimes(0);
        expect(rotateTo).toHaveBeenCalledTimes(0);
    });

    test('Test applyDefaultOptions with custom params', () => {
        applyDefaultOptions(cropper, {enable: false, zoomTo: 1, rotateTo: 90});
        expect(enable).toHaveBeenCalledTimes(0);
        expect(disable).toHaveBeenCalledTimes(1);
        expect(zoomTo).toHaveBeenCalledWith(1);
        expect(zoomTo).toHaveBeenCalledTimes(1);
        expect(rotateTo).toHaveBeenCalledTimes(1);
    });
});
