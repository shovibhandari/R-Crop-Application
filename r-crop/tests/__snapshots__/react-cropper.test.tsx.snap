// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`R crop render tests R crop snapshot 1`] = `
<div>
  <img
    alt="picture"
    src="https://cdn.pixabay.com/photo/2023/07/10/06/52/trail-8117575_1280.jpg"
    style={
      {
        "maxWidth": "100%",
        "opacity": 0,
      }
    }
  />
</div>
`;
